### ISSUE TITLE
Handle terminal borders with the w3m preview

### ISSUE THREAD

**Issue by LeLobster** [ranger/ranger#1332]
Continuation of what was talked about in PR #1327.

#### Reason
Some terminals configured with an inner border can give issues with the w3m preview being misaligned.  
Currently affected are `st` and `UXterm` (reported in issue #1314)  
While some terminals, like `urxvt`, correctly _set_ the available window size.  

#### Related issues
#1302 & #1110

#### Relevant information
`/ranger/ext/img_display.py _get_font_dimensions()`
can return _wrong_ values depending on whether `ws_xpixel` and `ws_ypixel` are set in `terminfo`

#### Todo
- Find a reliable way to calculate border space
- Figure out which terminals need it and which don't, and why
- Possibly change how ranger decides where to draw the w3m image

---
Honestly speaking, I don't think there will be a viable solution without a lot of _hacky_ stuff and hardcoded checks but I'll see if I can figure something out which would help.

**Comment by LeLobster** [comment]
[I've wrote this quick script to test some things](https://gist.github.com/LeLobster/942eb03d535c030423fdf725029251c4) but the results are not very helpful.  

Ran it in 4 different terminals with the exact same dimensions, `st, urxvt, xterm and terminator`.  
I figured getting the absolute window size, subtracting the (number of rows * font height ) from that and what's leftover would be the border top+bottom, but just getting the real window size is different with each terminal.  

`st`: `xwininfo` and `w3m` return the real size  
`urxvt` only `xwininfo` does  
`xterm`: all return the same value but the calculation is off (possibly because of the font size calculation)
`terminator`: values change depending on whether the statusbar/scrollbar is active

**Comment by LeLobster** [comment]
@toonn @Vifon Is there a specific reason `img_display.py` relies on `ioctl` for getting the font dimensions over `w3mimagedisplay`?

**Comment by toonn** [comment]
Re the first, I'm not sure. Although a more general method to retrieve information from a terminal doesn't seem like a bad thing to me.

Re the second, I think it's just slightly overclearing to be sure.

**Comment by toonn** [comment]
Also, note re the first that `w3mimagedisplay` *is* used as a fallback.

**Comment by noamcore** [comment]
Can I help in any way?

**Comment by toonn** [comment]
@noamcore, yes, figuring out any of the **Todo** bullet points @LeLobster  listed would definitely help.

**Comment by noamcore** [comment]
Analyzing the st's source-code, seems like the only way to get the internal padding is getting from the `WM_SIZE_HINTS`, which is `2 * borderpx`. I'm a little noob on python, but googling I've found that to achieve this, we need external packages.

The easiest way I've found is to patch the st/xst.

## xst version
```diff
diff --git i/src/xst.c w/src/xst.c
index 330ad44..fda4ec9 100644
--- i/src/xst.c
+++ w/src/xst.c
@@ -1697,8 +1697,8 @@ ttyresize(void)
 
 	w.ws_row = term.row;
 	w.ws_col = term.col;
-	w.ws_xpixel = xw.tw;
-	w.ws_ypixel = xw.th;
+	w.ws_xpixel = xw.tw + 2 * borderpx;
+	w.ws_ypixel = xw.th + 2 * borderpx;
 	if (ioctl(cmdfd, TIOCSWINSZ, &w) < 0)
 		fprintf(stderr, "Couldn't set window size: %s\n", strerror(errno));
 }
```

## original st
```diff
diff --git i/x.c w/x.c
index 0422421..9383f61 100644
--- i/x.c
+++ w/x.c
@@ -677,7 +677,7 @@ cresize(int width, int height)
 
 	tresize(col, row);
 	xresize(col, row);
-	ttyresize(win.tw, win.th);
+	ttyresize(win.tw + (2 * borderpx), win.th + (2 * borderpx));
 }
 
 void
```

Might be required to test with some another tools to see if this broke something up.

But for testing purpose, this fix the misalignment.

![xst](https://user-images.githubusercontent.com/44407744/50425517-ee50e780-0856-11e9-8578-1dbeac91ffd0.png)

**Comment by toonn** [comment]
Hmm, maybe this should be brought up wherever st issues are tracked, some suckless mailing list or something?

**Comment by noamcore** [comment]
Yeah, there is a mailing list for the suckess. I'll try to talk to them.

**Comment by seebye** [comment]
Offtopic: I just want to mention that `ueberzug` places the images correctly in stterm.  
Their implementation is way better as it allows in opposite to xterm to correctly measure the font size and padding.  
(By the way kitty implemented TIOCSWINSZ in the same way)  
So I hope they refuse to change it.

**Comment by toonn** [comment]
@seebye, bit of a bogus argument imo. They should provide both the window size and the available, i.e. borders excluded, size. Just because one of those fits your wheelhouse better doesn't mean it should be the only information available. Doesn't ueberzug require the border information for correct alignment too anyway? Imo having to calculate the font size is silly in the first place, would one word really matter in memory, maybe it used to at some point...

**Comment by seebye** [comment]
> Just because one of those fits your wheelhouse better doesn't mean it should be the only information available.

¯\_(ツ)_/¯

> They should provide both the window size and the available, i.e. borders excluded, size.

That's right, they should if there's a standard defined.  
Otherwise it won't be really useful as it will likely end like the implementation of displaying images in iterm, kitty, urxvt, etc..  
(A protocol per terminal emulator)

> Doesn't ueberzug require the border information for correct alignment too anyway?

No, it doesn't there's more information in the current implementation.  
If it doesn't output the size of the drawing area you can only guess the font size & padding so it's worse,  
but if it does you can calculate the exact font size and the padding.

> Imo having to calculate the font size is silly in the first place, would one word really matter in memory, maybe it used to at some point...

It's not about memory.  
I'm trying to make it as easy as possible to display images in terminals,
that's why ueberzug uses characters as size units which have to be converted to pixels that's what the font size is used to do.  
It's easier to place them at the correct position if you don't have to worry about pixels (as everything else is measured in chars).

**Comment by toonn** [comment]
I mean it's silly terminals don't report their font size. I have nothing against what ueberzug does. The drawing area *is* the area inside the borders though, I don't really understand why ueberzug needs the dimensions of the st window instead of the ones inside the borders.

Ueberzug is great, it looks like we might recommend it over w3m when it's in the next release, it's not a panacea though. It requires X which is a pretty big assumption and makes it non-crossplatform.

**Comment by seebye** [comment]
> The drawing area is the area inside the borders though

Sorry, I meant the drawing area the text is placed in.  

> I don't really understand why ueberzug needs the dimensions of the st window instead of the ones inside the borders.

It doesn't need them it just notices that it can calculate the correct padding & font size.  

It's simple:
- (a) st returns the dimension of the text area
- (b) X11 returns the window size

padding = (b) - (a)
font size = (a) / columns (or (a) / rows for height)

**Comment by toonn** [comment]
Ah, so it's not cross platform? We have to pretend the terminal's the only source of information.

**Comment by LeLobster** [comment]
Nice to see the activity in this issue, sadly I haven't been able to make any substantial progress due to some serious personal issues. Things are settling down now though so I'll try to find some time as soon as possible.  

All progress I had made so far though are in my `crossterm-borderfix` branch @ https://github.com/LeLobster/ranger/tree/crossterm-borderfix

If anyone wants to take a look, feel free to do so.  It was the most consistent way of calculating I could think of at the time I was experimenting around, still not 100% reliable but fairly close.
